lib/commons/dom/url-props-from-attribute.js

```diff
@@ -0,0 +1,154 @@
+ * @returns {Object}
+ */
+dom.urlPropsFromAttribute = function urlPropsFromAttribute(node, attribute) {
+	const value = node[attribute];
+	if (!value) {
+		return undefined;
+	}
+
+	const nodeName = node.nodeName.toUpperCase();
+	let parser = node;
+
+	/**
+	 * Note:
+	 * The need to create a parser, is to keep this function generic, to be able to parse resource from element like `iframe` with `src` attribute
+	 */
+	if (!['A', 'AREA'].includes(nodeName)) {
+		parser = document.createElement('a');
+		parser.href = value;
+	}
+
+	const { pathname = '', filename = '' } = getPathnameOrFilename(
+		parser.pathname
+	);
+	return {
+		protocol: parser.protocol,
+		hostname: parser.hostname,
+		port: getPort(parser.port),
+		pathname: pathname.replace(/\/$/, ''), // remove trialing slashes
+		search: getSearchPairs(parser.search),
+		hash: getHashRoute(parser.hash),
+		filename
+	};
+};
+
+
+/**
+ * Resolve given port excluding default port(s)
+ * @param {String} port port
+ * @returns {String}
+ */
+function getPort(port) {
+	/**
+	 * Note: IE11 resolves port to `443` for any URL with `https:`
+	 */
+	const excludePorts = [`443`, `80`]
+	return !excludePorts.includes(port)
+		? port
+		: ``
+}
+
+/**
+ * Resolve if a given pathname has filename & resolve the same as parts
+ * @method getPathnameOrFilename
+ * @param {String} pathname pathname part of a given uri
+ * @returns {Array<String>}
+ */
+function getPathnameOrFilename(pathname) {
+	if (!pathname) {
+		return;
+	}
+
+	const filename = pathname.split('/').pop();
+	if (!filename || filename.indexOf('.') === -1) {
+		return {
+			pathname
+		};
+	}
+
+	return {
+		// remove `filename` from `pathname`
+		pathname: pathname.replace(filename, ''),
+
+		// ignore filename when index.*
+		filename: /index./.test(filename)
+			? undefined
+			: filename
+	};
+}
+
+/**
+ * Parse a given query string to key/value pairs sorted alphabetically
+ * @param {String} searchStr search string
+ * @returns {Object}
+ */
+function getSearchPairs(searchStr) {
+	const query = {};
+
+	if (!searchStr || !searchStr.length) {
+		return query
+	}
+
+	const curatedSearchStr = searchStr[0] === `?` ? searchStr.substring(1) : searchStr
+	const pairs = curatedSearchStr.split(`&`);
+	if (!pairs || !pairs.length) {
+		return query
+	}
+
+
+	for (let index = 0; index < pairs.length; index++) {
+		const pair = pairs[index];
+		const [key, value = ''] = pair.split(`=`)
+		query[decodeURIComponent(key)] = decodeURIComponent(value);
+	}
+
+	/**
+	 * Sort query `keys` alphabetically
+	 */
+	return Object.keys(query)
+		.sort()
+		.reduce((out, key) => {
+			out[key] = query[key]
+			return out
+		}, {});
+}
+
+/**
+ * Interpret a given hash
+ * if `hash`
+ * -> is `hashbang` -or- `hash` is followed by `slash`
+ * -> it resolves to a different resource
+ * @method getHashRoute
+ * @param {String} hash hash component of a parsed uri
+ * @returns {String|undefined}
```

Comment:
These methods are inconsistent about what happens when they didn't find anything. Some return undefined, other return an empty string. I don't really care which one is used, but it should be consistent.